api/src/routes/challenge.ts

```diff
@@ -424,5 +426,104 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.post(
+    '/ms-trophy-challenge-completed',
+    {
+      schema: schemas.msTrophyChallengeCompleted,
+      errorHandler(error, request, reply) {
+        if (error.validation) {
+          void reply.code(400);
+          void reply.send({ type: 'error', message: 'flash.ms.trophy.err-2' });
+        } else {
+          fastify.errorHandler(error, request, reply);
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        const challengeId = req.body.id;
+        const challenge = msTrophyChallenges.find(
+          challenge => challenge.id === challengeId
+        );
+
+        if (!challenge) {
+          return reply
+            .code(400)
+            .send({ type: 'error', message: 'flash.ms.trophy.err-2' });
+        }
+
+        const msUser = await fastify.prisma.msUsername.findFirst({
+          where: { userId: req.session.user.id }
+        });
+
+        if (!msUser || !msUser.msUsername) {
+          return reply
+            .code(403)
+            .send({ type: 'error', message: 'flash.ms.trophy.err-1' });
+        }
+
+        const { msUsername } = msUser;
+
+        // TODO: log error if msTrophyId not found?
+        const msTrophyId = challenge.msTrophyId ?? '';
+
+        const msTrophyStatus = await verifyTrophyWithMicrosoft({
+          msUsername,
+          msTrophyId
+        });
+
+        if (msTrophyStatus.type === 'error') {
+          return reply.code(403).send(msTrophyStatus);
+        }
+
+        const user = await fastify.prisma.user.findUniqueOrThrow({
+          where: { id: req.session.user.id },
+          select: { completedChallenges: true, progressTimestamps: true }
+        });
+
+        const { completedChallenges } = user;
```

Comment:
Curious why we didn't use the `updateUserChallengeData` function here

Reply:
That function has to handle too many special cases and is quite confusing. I'd rather be a little less DRY for now, while we're building the functionality, and try to refactor later if needed.